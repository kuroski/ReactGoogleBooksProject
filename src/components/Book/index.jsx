import React, {Component} from 'react'
import {Link} from 'react-router'
import Highlighter from 'react-highlight-words'

const propTypes = {
  bookId: React.PropTypes.string.isRequired,
  index: React.PropTypes.number.isRequired,
  title: React.PropTypes.string.isRequired,
  onFavorite: React.PropTypes.func.isRequired,
  isOnFavorite: React.PropTypes.func.isRequired,
  term: React.PropTypes.string
}

class Book extends Component {
  constructor(props) {
    super(props)
    this.favoriteBook = this.favoriteBook.bind(this)
    this.isOnFavorite = this.isOnFavorite.bind(this)
    this.renderFavoritedBook = this.renderFavoritedBook.bind(this)
    this.renderBook = this.renderBook.bind(this)
  }

  isOnFavorite(bookId) {
    return this.props.isOnFavorite(bookId)
  }

  favoriteBook(bookId, index) {
    this.props.onFavorite(bookId, index)
  }

  renderFavoritedBook() {
    return (
      <li className="c-book c-book--favorited" key={this.props.index}>
        <Highlighter
          searchWords={[this.props.term]}
          textToHighlight={this.props.title}
          highlightClassName="Highlight"
        />
        <Link to={`/${this.props.bookId}`}>Detail</Link>
      </li>
    )
  }

  renderBook(item, index) {
    return (
      <li className="c-book" key={this.props.index}>
        <Highlighter
          searchWords={[this.props.term]}
          textToHighlight={this.props.title}
          highlightClassName="Highlight"
        />
        <Link to={`/${this.props.bookId}`}>Detail</Link>
        <button onClick={() => this.favoriteBook(this.props.bookId, this.props.index)} className='c-book__favorite'>
          Favorite
        </button>
      </li>
    )
  }

  render() {
    if(this.isOnFavorite(this.props.bookId)) {
      return this.renderFavoritedBook()
    }
    return this.renderBook()
  }
}

Book.propTypes = propTypes

export default Book