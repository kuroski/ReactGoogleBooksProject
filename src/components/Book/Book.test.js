import React from 'react'
import Book from './'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import booksJson from '../../test/__mocks__/books.json'

describe('Book', () => {
  it('renders correctly', () => {
    const BookTree = renderer.create(
      <Book bookId="001" index={0} title="Harry Potter" onFavorite={() => ''} isOnFavorite={() => ''} />
    ).toJSON()
    expect(BookTree).toMatchSnapshot()
  })

  it('should call onFavoriteBook when favorite button is clicked', () => {
    const onFavoriteBookSpy = jest.fn()
    const firstBook = booksJson.items[0]
    const wrapper = shallow(
      <Book
        bookId={firstBook.id}
        index={0}
        title={firstBook.volumeInfo.title}
        onFavorite={onFavoriteBookSpy}
        isOnFavorite={() => ''} />)
    const firstFavoriteButton = wrapper.find('.c-book__favorite').first()

    firstFavoriteButton.simulate('click', firstBook.id, 0)

    expect(onFavoriteBookSpy).toHaveBeenCalledTimes(1)
    expect(onFavoriteBookSpy).toHaveBeenCalledWith(firstBook.id, 0)
  })

  it('should check if the book is favorited', () => {
    const isOnFavorite = jest.fn()
    const wrapper = shallow(<Book bookId="001" index={0} title="Harry Potter" onFavorite={() => ''} isOnFavorite={isOnFavorite} />)

    expect(isOnFavorite).toHaveBeenCalledTimes(1)
    expect(isOnFavorite).toHaveBeenLastCalledWith(wrapper.instance().props.bookId)
  })

  it('should hide the favorite button when the book is in favorites list', () => {
    const isOnFavorite = jest.fn().mockReturnValueOnce(true)
    const wrapper = shallow(<Book bookId="001" index={0} title="Harry Potter" onFavorite={() => ''} isOnFavorite={isOnFavorite} />)

    expect(wrapper.hasClass('c-book--favorited')).toBe(true)
    expect(wrapper.find('.c-book__favorite')).toHaveLength(0)
  })
})