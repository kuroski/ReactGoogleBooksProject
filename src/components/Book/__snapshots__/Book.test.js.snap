exports[`Book renders correctly 1`] = `
<li
  className="c-book">
  <span>
    <span>
      Harry Potter
    </span>
  </span>
  <a
    onClick={[Function]}
    style={Object {}}>
    Detail
  </a>
  <button
    className="c-book__favorite"
    onClick={[Function]}>
    Favorite
  </button>
</li>
`;
